python/ray/ml/utils/huggingface_utils.py

```diff
@@ -0,0 +1,94 @@
+from pathlib import Path
+from typing import Dict, Generator, Iterator, List
+
+import torch
+import transformers.trainer
+from torch.utils.data import IterableDataset
+from transformers.trainer_callback import TrainerCallback
+
+from ray import train
+from ray.util import get_node_ip_address
+from ray.data.dataset import Dataset
+
+CHECKPOINT_PATH_ON_NODE_KEY = "checkpoint_path_on_node"
+NODE_IP_KEY = "node_ip"
+
+
+class HFIterableDatasetWithLen(IterableDataset):
+    """Special Torch IterableDataset with preset length."""
+
+    def __init__(self, generator: Generator, length: int):
+        self.generator = generator
+        self._len = length
+
+    def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
+        it = self.generator
+        for x in it:
+            # HF-specific format
+            yield {**x[0], "labels": x[1]}
```

Comment:
Can you add a comment explaining why we need this? I could not find anything about this format in the Huggingface documentation.

Reply:
This has been determined by analysis of the HF code. Indeed, using an arbitrary format leads to exceptions on part of huggingface.

Reply:
Got it- it might be better to just match the behavior of Huggingface.

Reply:
`to_torch` already returns a format that is not compatible, so we have to convert it somewhere.

Reply:
Hmm ok got it. My concern is that if the format that they expect is implicit, then it can be arbitrarily changed in the future, and we would have to keep updating our code to match it.

This might good to ask the HF team, to see if there is a publicly documented format they expect.